rust/core-lib/src/find.rs

```diff
@@ -263,29 +263,39 @@ impl Find {
     /// Return the occurrence closest to the provided selection `sel`. If searched is reversed then
     /// the occurrence closest to the start of the selection is returned. `wrapped` indicates that
     /// if the end of the text is reached the search continues from the start.
-    pub fn next_occurrence(&self, text: &Rope, reverse: bool, wrapped: bool, sel: (usize, usize)) -> Option<SelRegion> {
-        let (sel_start, sel_end) = sel;
-
+    pub fn next_occurrence(&self, text: &Rope, reverse: bool, wrapped: bool, sel: &Selection) -> Option<SelRegion> {
         if self.occurrences.len() == 0 {
             return None;
         }
 
+        // last selection is used to find next occurrence
+        let (sel_start, sel_end) = match sel.last() {
+            Some(last) => (last.min(), last.max()),
+            None => (0, 0)
+        };
+
         if reverse {
             let next_occurrence = match sel_start.checked_sub(1) {
                 Some(search_end) => self.occurrences.regions_in_range(0, search_end).last(),
                 None => None
             };
 
             if next_occurrence.is_none() && !wrapped {
-                return self.occurrences.regions_in_range(0, text.len()).last().cloned();
+                // get last unselected occurrence
+                return self.occurrences.regions_in_range(0, text.len()).iter().cloned().filter(|o| {
+                    sel.regions_in_range(o.min(), o.max()).is_empty()
+                }).collect::<Vec<SelRegion>>().last().cloned();
             }
 
             next_occurrence.cloned()
         } else {
             let next_occurrence = self.occurrences.regions_in_range(sel_end + 1, text.len()).first();
 
             if next_occurrence.is_none() && !wrapped {
-                return self.occurrences.regions_in_range(0, text.len()).first().cloned();
+                // get first unselected occurrence
+                return self.occurrences.regions_in_range(0, text.len()).iter().cloned().filter(|o| {
+                    sel.regions_in_range(o.min(), o.max()).is_empty()
+                }).collect::<Vec<SelRegion>>().first().cloned();
```

Comment:
I think `.next().cloned()` is a more efficient approach than collecting. However, that's not as easy to do for the last, and this code is more parallel to the stanza above, so I'm not suggesting changing it.